File: packages/shared/schemas/attendance.ts
import { z } from "zod";

export const CreateAttendanceSchema = z.object({
  eventId: z.string(),
  event: z.unknown(),
  userId: z.string(),
  user: z.unknown(),
  status: z.string(),
  selfieUrl: z.string(),
  signatureUrl: z.string(),
  checkedInAt: z.string().nullable(),
  version: z.number().int(),
});

export const UpdateAttendanceSchema = z.object({
  eventId: z.string().optional(),
  event: z.unknown().optional(),
  userId: z.string().optional(),
  user: z.unknown().optional(),
  status: z.string().optional(),
  selfieUrl: z.string().optional(),
  signatureUrl: z.string().optional(),
  checkedInAt: z.string().nullable(),
  version: z.number().int().optional(),
});

export type CreateAttendanceInput = z.infer<typeof CreateAttendanceSchema>;
export type UpdateAttendanceInput = z.infer<typeof UpdateAttendanceSchema>;

File: apps/api/internal/models/attendance.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Attendance represents a attendance in the system.
type Attendance struct {
	ID        string         `gorm:"primarykey;size:36" json:"id"`
	EventID string `gorm:"size:36;uniqueIndex:idx_attendance_event_user" json:"event_id" binding:"required"`
	Event Event `gorm:"foreignKey:EventID" json:"event"`
	UserID string `gorm:"size:36;uniqueIndex:idx_attendance_event_user" json:"user_id" binding:"required"`
	User User `gorm:"foreignKey:UserID" json:"user"`
	Status string `gorm:"size:255" json:"status" binding:"required"`
	SelfieUrl string `gorm:"size:255" json:"selfie_url"`
	SignatureUrl string `gorm:"size:255" json:"signature_url"`
	CheckedInAt *time.Time `json:"checked_in_at"`
	Version   int            `gorm:"not null;default:1" json:"version"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// BeforeCreate generates a UUID before inserting.
func (m *Attendance) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = uuid.New().String()
	}
	return nil
}

// BeforeUpdate increments Version so offline clients can detect server-side updates.
func (m *Attendance) BeforeUpdate(tx *gorm.DB) error {
	m.Version++
	return nil
}
